### DISCUSSION TITLE
Android Nearby Share compatibility with Windows Nearby Connections

### DISCUSSION THREAD

**Discussion by rajendrasp** [google/nearby#2446] [General]
### Project

Nearby Connections

### Language

C++

### OS Platform

Windows

### What happened?

I have setup the Nearby Connections library on Windows. It is building fine and I am able to StartDiscovery. However in the logs, it gives me following message when it finds my phone - 

I0129 00:54:09.772869 46244 p2p_cluster_pcp_handler.cc:224] OnePlus Nord doesn't have any endpoint id, discarding.
I0129 00:54:09.772869 46244 p2p_cluster_pcp_handler.cc:274] Found unrecognized BluetoothDeviceName OnePlus Nord

I am trying with Nearby share from Android and why does it not have an endpoint id?
Would I be able to integrate the sample app created on Windows with Nearby share from Android?

### What did you expect to happen?

My phone to get discovered on my Windows app which is using Nearby Connections library.

### How can we reproduce it (as minimally and precisely as possible)?

I am trying with integrating the Connections library in a Windows app.

### How often does this bug happen?

Every time

**Comment by bourdakos1** [comment]
Hey @rajendrasp, glad to hear you were able to get Nearby Connections building/running on Windows, that’s awesome!

In order to discover Android devices running Nearby Share on Windows, you’d need to use the same service ID for discovery as Nearby Share. I’m not sure what service ID Nearby Share uses, but if you search the ChromeOS codebase you should be able to find it. Additionally, the Android device will need to be in the “everyone” discoverability mode.

I should mention that even if you discover Android devices with Nearby Connections, it will be difficult to send data. Since Nearby Share uses a custom packet format, you won’t simply be able to send files. I recommend making your own app that uses Nearby Connections on Android if you need to be able to also send data.

As a side note, would you be interested in contributing your work to get Windows building? I think it would help a lot of other people 😊

**Comment by rajendrasp** [comment]
Thank you @bourdakos1 for your response. I would love to contribute my work to get this building on Windows. It was a long trail of errors to fix. And I agree it would help many others.
But I have few items to nail down.
I would like to understand more about getting my Windows app to work with Google Nearby Share that comes out of the box on Android devices.
You mention ChromeOS. Is this the code you are talking about? Any pointers there would help. https://source.chromium.org/chromium/chromium/src/+/main:chrome/services/sharing/nearby/nearby_connections.cc
You also mention that Neaby Share uses a custom packet format. Any more info on that? Is that open source?

**Comment by bourdakos1** [comment]
Looks like the Nearby Share source is also being added to this repo. You can find the service ID here: https://github.com/google/nearby/blob/9be93ae23dd3d1a107bb4c0684aea6503f89ba0d/sharing/nearby_connections_manager_impl.cc#L58

**Comment by rajendrasp** [comment]
Thanks for pointing to this.
I tried with this service id but no luck.
When I am using this service id, I notice that there are peripheral devices getting discovered. For example, I see following log - 

I0131 16:54:12.954299 38132 client_proxy.cc:330] ClientProxy [Endpoint Found]: [enter] id=E6AS; service=NearbySharing; info=325cc78f0e172a8d2faaf1373ddcb03a70; medium=BLE
Found endpoint E6AS on service NearbySharing

But for my phone, follwing is the data discovered - ([code location](https://github.com/google/nearby/blob/80c0d0916e8e5c3637b389a14c4cc45ee30bb8ce/connections/implementation/p2p_cluster_pcp_handler.cc#L283))

![image](https://github.com/google/nearby/assets/51284446/912f2c6f-a867-4b1e-b95a-56d8843d5774)

It also discovered another device with follwoing info - 

![image](https://github.com/google/nearby/assets/51284446/e051e89f-4c86-4fcb-a4b0-c83c8006eb62)

Is that the issue in Windows implementation or the advertising part?

**Comment by bourdakos1** [comment]
It sounds like it was able to discover a Nearby Share device, but the "OnePlus Nord" bluetooth advertisement in your screenshot doesn't look to be an advertisement from Nearby Share

**Comment by rajendrasp** [comment]
Is there a Android version requirement on Phone for Nearby share? Mine is Android 12. 
So if this is not Nearby Share advertisement, then it might be normal Bluetooth advertisement from phone.

**Comment by rajendrasp** [comment]
I tried capturing the bluetooth traffic incoming from the phone [using Wireshark on PC](https://bleak.readthedocs.io/en/latest/troubleshooting.html#windows-10) . It does not show NearbyShare advertisement. All the packets were like below - 
Nearby Share will use "Bluetooth address" to advertise right?
What is the missing piece here?

Bluetooth HCI Event - Extended Inquiry Result
    Event Code: Extended Inquiry Result (0x2f)
    Parameter Total Length: 255
    Number of responses: 1
    BD_ADDR: OnePlusTechn_38:eb:e9 (ac:5f:ea:38:eb:e9)
    Page Scan Repetition Mode: R1 (0x01)
    Reserved: 0x00
    Class of Device: 0x5a020c (Phone:Smartphone - services: Networking Capturing ObjectTransfer Telephony)
        0000 11.. = Minor Device Class: Smartphone (0x03)
        .... ..00 = Format Type: 0x0
        0... .... .... .... = Major Service Classes: Information: False
        .1.. .... .... .... = Major Service Classes: Telephony: True
        ..0. .... .... .... = Major Service Classes: Audio: False
        ...1 .... .... .... = Major Service Classes: Object Transfer: True
        .... 1... .... .... = Major Service Classes: Capturing: True
        .... .0.. .... .... = Major Service Classes: Rendering: False
        .... ..1. .... .... = Major Service Classes: Networking: True
        .... ...0 .... .... = Major Service Classes: Positioning: False
        .... .... 0... .... = Major Service Classes: Reserved: 0x0
        .... .... .0.. .... = Major Service Classes: LE Audio: 0x0
        .... .... ..0. .... = Major Service Classes: Limited Discoverable Mode: False
        .... .... ...0 0010 = Major Device Class: Phone (0x02)
    .000 0111 0010 1100 = Clock Offset: 0x072c
    RSSI: -62 dBm
    Extended Inquiry Response Data
        Device Name: OnePlus Nord
            Length: 13
            Type: Device Name (0x09)
            Device Name: OnePlus Nord
        16-bit Service Class UUIDs
            Length: 23
            Type: 16-bit Service Class UUIDs (0x03)
            UUID 16: OBEX Object Push (0x1105)
            UUID 16: Audio Source (0x110a)
            UUID 16: A/V Remote Control Target (0x110c)
            UUID 16: A/V Remote Control (0x110e)
            UUID 16: Headset - Audio Gateway (0x1112)
            UUID 16: PANU (0x1115)
            UUID 16: NAP (0x1116)
            UUID 16: Handsfree Audio Gateway (0x111f)
            UUID 16: Phonebook Access - PSE (0x112f)
            UUID 16: PnP Information (0x1200)
            UUID 16: Message Access Server (0x1132)
        32-bit Service Class UUIDs
            Length: 1
            Type: 32-bit Service Class UUIDs (0x05)
        128-bit Service Class UUIDs
            Length: 129
            Type: 128-bit Service Class UUIDs (0x07)
            Custom UUID: 00000000-0000-0000-0000-000000000000 (Unknown)
            Custom UUID: 00000000-0000-0000-0000-000000000000 (Unknown)
            Custom UUID: 00000000-0000-0000-0000-000000000000 (Unknown)
            Custom UUID: 00000000-0000-0000-0000-000000000000 (Unknown)
            Custom UUID: 00000000-0000-0000-0000-000000000000 (Unknown)
            Custom UUID: 00000000-0000-0000-0000-000000000000 (Unknown)
            Custom UUID: 00000000-0000-0000-0000-000000000000 (Unknown)
            Custom UUID: 00000000-0000-0000-0000-000000000000 (Unknown)
        Unused

**Comment by bourdakos1** [comment]
You can search for "Nearby Share" in the setting of your phone. Make sure Nearby Share is turned on and has its visibility set to "everyone"

**Comment by rajendrasp** [comment]
ok, so phone is discovered and connection accepted by phone from my console app when Google Nearby share is running on my windows pc where console app is running. Phone sends a disconnect message when connection is established. And payload transfer does not happen. I am attaching the log below - 
When tried to send file from Phone to app, there also it connects and starts the transfer. But disconnect after getting first packet.

[logss.txt](https://github.com/google/nearby/files/14166990/logss.txt)

**Comment by bourdakos1** [comment]
Oh yes, I forgot to mention, your phone will need to be "nudged" into a discoverable mode. This can either be done by starting discovery on another device (in this case it was your Windows PC) or by using the "Files by Google" app and clicking "Receive" in the "Nearby Share" section of the app.

Nearby Share will disconnect if you send payloads that don't follow Nearby Share's custom packet format.

**Comment by rajendrasp** [comment]
thanks. Just to confirm since this will mostly need building ./sharing folder for windows as well which is bit of work.
are you talking about following custom format - 

https://github.com/google/nearby/blob/6ad4bf75e2e917571b42201fa2b5f30f8ba534db/sharing/proto/wire_format.proto#L134

https://github.com/google/nearby/blob/6ad4bf75e2e917571b42201fa2b5f30f8ba534db/sharing/nearby_sharing_decoder_impl.cc#L35

Going through the code, I see that it sends an introduction frame first and then the payload. I will need to match what it looks for in the receiving part of code. do you have any documentation on the ./sharing code?

**Comment by rajendrasp** [comment]
Hi, I am trying to figure out the custom packet format and it is requiring me to add a lot of dependent code from ./sharing folder to my project. Right now, I am stuck at resolving the dependency of PublicCertificateDatabase.

https://github.com/google/nearby/blob/f8c907a1201f9c17ea0e36570d842da89fc55131/sharing/internal/api/public_certificate_database.h#L28

1. Can I skip paired key verification on both mobile and my console app? 
2. If not, how do I resolve PublicCertificateDatabase dependency? I did not see this in Chromium codebase as well. In general, I did not find the implementation of anything inside \sharing\internal\api. Is that code not yet added to the repo?

**Comment by bourdakos1** [comment]
Yes, you can skip the certificate stuff. That's just used for contacts/self share.

**Comment by rajendrasp** [comment]
Looking at the code, looks like it needs certificate stuff - 

https://github.com/google/nearby/blob/f8c907a1201f9c17ea0e36570d842da89fc55131/sharing/nearby_sharing_service_impl.cc#L3213

https://github.com/google/nearby/blob/f8c907a1201f9c17ea0e36570d842da89fc55131/sharing/nearby_sharing_service_impl.cc#L3431

How do I skip it? If I own both apps, I can skip the flow but I need to communicate with Google's Nearby share on Android device.

**Comment by rajendrasp** [comment]
This part of code also does not make sense - 

https://github.com/google/nearby/blob/f8c907a1201f9c17ea0e36570d842da89fc55131/sharing/paired_key_verification_runner.cc#L309

There is no action on receiving end when cert info is received - 

https://github.com/google/nearby/blob/f8c907a1201f9c17ea0e36570d842da89fc55131/sharing/nearby_sharing_service_impl.cc#L4048

How can I send my public cert to my mobile device and have Nearby Share use it?

It says that it is to be shared offline. How is that shared offline?

https://github.com/google/nearby/blob/f8c907a1201f9c17ea0e36570d842da89fc55131/sharing/proto/wire_format.proto#L295

**Comment by bourdakos1** [comment]
The certificate part isn't open source. You won't be able to build ./sharing without it, but if you manually write/parse frames, it can be skipped. If you don't use certificates, your app will just function as "everyone" mode in Nearby Share. Other devices will always be able to discover your app, but you will only be able to discover NS devices if they set their visibility to "everyone".

**Comment by rajendrasp** [comment]
Thank you @bourdakos1 ! I am able to successfully transfer the file to phone from my app. Really appreciate your help in resolving the issues along the way!
One question - when I use Google Quick share Windows app, I do not need to go to Files app on Android and start Receive part. Can I achieve that from my Windows app? How can my phone be "nudged" into a discoverable mode without having to open Files app or any other app on Phone?

**Comment by bourdakos1** [comment]
Yay!!

You can use a fast initiation advertisement to nudge the other device: https://github.com/google/nearby/blob/88064b1543b807d5c3de1bb729058848d301cf17/sharing/fast_initiation/nearby_fast_initiation_impl.cc#L107

**Comment by rajendrasp** [comment]
I see that fast initiation implementation is also missing in ./sharing folder. E.g. FastInitiationManager implementation is not available. 
I saw fast initiation advertiser implementation here - https://chromium.googlesource.com/chromium/src/+/refs/heads/main/chrome/browser/nearby_sharing/fast_initiation/fast_initiation_advertiser.cc

which has dependency on adapter which looks like this - 

https://source.chromium.org/chromium/chromium/src/+/main:device/bluetooth/bluetooth_adapter_winrt.h

Do you think this is what would be needed for fast initiation on Windows?

Also, to confirm, it is the fast initiation **advertisement** from the sender which nudges the other device; is that correct? Can you share any details how that works?

**Comment by bourdakos1** [comment]
Windows only needs the fast init advertiser to function, and correct, the advertisement is from the sender which nudges the other device.

**Comment by rajendrasp** [comment]
I came across this documentation of this library. https://github.com/grishka/NearDrop/blob/master/PROTOCOL.md
Do you know if this is accurate? It would help to have something official by the authors of the library.

**Comment by bourdakos1** [comment]
Looks semi correct. However, I don’t recall the exact advertisement format off the top of my head.

**Comment by aditya-mankal** [comment]
Hello @bourdakos1, hope you are doing well! @rajendrasp and I have worked on setting up Google Nearby on Windows. While he is occupied elsewhere, I will be working on the fast initiation advertiser. 

Can you please tell help me with exact advertisement format? My plan is to broadcast it using a quick piece of code before investing in integrating Fast Initiation into our current fork of Google Nearby. 

Thank you!

**Comment by aditya-mankal** [comment]
Hello @bourdakos1, hope you are doing well!
Looking forward to your inputs on fast initiation. Knowing the exact BLE broadcast to nudge the Android devices into receive mode will greatly improve the user experience for my project. Thank you!

**Comment by bourdakos1** [comment]
I think you should be able to hardcode the advertisement to these values:
- Service UUID: `0xFE2C`
- Service data: `0xFC128E0042`
